Makefile

```diff
@@ -1,25 +1,30 @@
 CPP          = g++
-CPPOPTS      = -Wall -std=c++11 -Iinclude/
+CPPOPTS      = -Wall -Werror -std=c++11 -Iinclude/
 DEBUG_OPTS   = -fprofile-arcs -ftest-coverage -O0 -g -fPIC
 RELEASE_OPTS = -O3
+DEBUG_LIBS   = debug/url.o
+RELEASE_LIBS = release/url.o
 
-release: release/url.o
+release: $(RELEASE_LIBS)
 
-release/url.o: src/url.cpp include/url.h
+release/%.o: src/%.cpp include/%.h
 	mkdir -p release
 	$(CPP) $(CPPOPTS) $(RELEASE_OPTS) -o $@ -c $<
 
-debug: debug/url.o
+debug: $(DEBUG_LIBS)
 
-debug/url.o: src/url.cpp include/url.h
+debug/%.o: src/%.cpp include/%.h
 	mkdir -p debug
 	$(CPP) $(CPPOPTS) $(DEBUG_OPTS) -o $@ -c $<
 
 bench: bench.cpp release
-	$(CPP) $(CPPOPTS) $(RELEASE_OPTS) -o $@ $< release/url.o
+	$(CPP) $(CPPOPTS) $(RELEASE_OPTS) -o $@ $< $(RELEASE_LIBS)
 
-run-tests: test.cpp debug
-	$(CPP) $(CPPOPTS) $(DEBUG_OPTS) -o $@ $< debug/url.o -lgtest -lpthread
+test/%.o: test/%.cpp
+	$(CPP) $(CPPOPTS) $(DEBUG_OPTS) -o $@ -c $<
+
+run-tests: test/test-all.o test/test-url.o $(DEBUG_LIBS)
+	$(CPP) $(CPPOPTS) $(DEBUG_OPTS) -o $@ $^ -lgtest -lpthread
```

Comment:
I would call the test binary `test-all` since that's the object it is generated from and where the `main` originates. Alternatively, I would rename `test-all.c` to be `run-tests.c` then it would be clearer.